### PR TITLE
Improve debugger StackOverflow handling

### LINKED ISSUE #70755: Crash in coreclr.dll!Debugger::SendException while processing stack overflow in x64 Windows
### Description

The in-process part of the .NET debugging services will exhaust the guard page while trying to send a stack overflow exception. This results in the target process exiting with an access violation and the debugger never breaking on the stack overflow. I don't know if this is the only way to cause this or just the easiest way, but I don't see this with a "hello world" sync method, but I do see it with an async method.

### Reproduction Steps

Run the following program under a debugger in x64 Windows. This repros with either 6.0.5 or with .NET 7 preview 4.

```csharp
class Program
{
    public static async Task Main(string[] args)
    {
        await Task.Delay(100);
        string result = StackOverflow(100000);
        Console.WriteLine(result);
    }

    private static string StackOverflow(int v)
    {
        if (v == 0)
        {
            return string.Empty;
        }
        else
        {
            return "*" + StackOverflow(v - 1);
        }
    }
}
```

### Expected behavior

Debugger should receive an unhandled exception event for the stack overflow.

### Actual behavior

Target process crashes with the following stack:

```
void StackFrameIterator::ProcessIp(PCODE Ip)
{
...
    // Re-initialize codeInfo with new IP
    m_crawl.codeInfo.Init(Ip, m_scanFlag);
   00007FFFF709840D  add         rcx,230h  
   00007FFFF7098414  mov         r8d,dword ptr [rbx+2E0h]  
=> 00007FFFF709841B  call        EECodeInfo::Init (07FFFF7098438h)  
```

The `call` instruction crashes because rsp already points at the end of stack, so the following page in unmapped.

```
coreclr.dll!StackFrameIterator::ProcessIp(unsigned __int64 Ip) Line 2729
	at D:\a\_work\1\s\src\coreclr\vm\stackwalk.cpp(2729)
coreclr.dll!StackFrameIterator::Init(Thread * pThread, Frame * pFrame, REGDISPLAY * pRegDisp, unsigned int flags) Line 1211
	at D:\a\_work\1\s\src\coreclr\vm\stackwalk.cpp(1211)
coreclr.dll!Thread::StackWalkFramesEx(REGDISPLAY * pRD, StackWalkAction(*)(CrawlFrame *, void *) pCallback, void * pData, unsigned int flags, Frame * pStartFrame) Line 907
	at D:\a\_work\1\s\src\coreclr\vm\stackwalk.cpp(907)
coreclr.dll!EEDbgInterfaceImpl::StackWalkFramesEx(Thread * pThread, REGDISPLAY * pRD, StackWalkAction(*)(CrawlFrame *, void *) pCallback, void * pData, unsigned int flags) Line 78
	at D:\a\_work\1\s\src\coreclr\vm\eedbginterfaceimpl.cpp(78)
coreclr.dll!Debugger::IsThreadAtSafePlaceWorker(Thread * thread) Line 12563
	at D:\a\_work\1\s\src\coreclr\debug\ee\debugger.cpp(12563)
coreclr.dll!AtSafePlaceHolder::AtSafePlaceHolder(Thread * pThread) Line 195
	at D:\a\_work\1\s\src\coreclr\debug\ee\debugger.cpp(195)
coreclr.dll!Debugger::SendException(Thread * pThread, bool fFirstChance, unsigned __int64 currentIP, unsigned __int64 currentSP, bool fContinuable, bool fAttaching, bool fForceNonInterceptable, _EXCEPTION_POINTERS * pExceptionInfo) Line 7690
	at D:\a\_work\1\s\src\coreclr\debug\ee\debugger.cpp(7690)
coreclr.dll!Debugger::LastChanceManagedException(_EXCEPTION_POINTERS * pExceptionInfo, Thread * pThread, int jitAttachRequested) Line 8656
	at D:\a\_work\1\s\src\coreclr\debug\ee\debugger.cpp(8656)
coreclr.dll!NotifyDebuggerLastChance(Thread * pThread, _EXCEPTION_POINTERS * pExceptionInfo, int jitAttachRequested) Line 3603
	at D:\a\_work\1\s\src\coreclr\vm\excep.cpp(3603)
coreclr.dll!WatsonLastChance(Thread * pThread, _EXCEPTION_POINTERS * pExceptionInfo, TypeOfReportedError tore) Line 3994
	at D:\a\_work\1\s\src\coreclr\vm\excep.cpp(3994)
coreclr.dll!EEPolicy::HandleFatalStackOverflow(_EXCEPTION_POINTERS * pExceptionInfo, int) Line 725
	at D:\a\_work\1\s\src\coreclr\vm\eepolicy.cpp(725)
coreclr.dll!EEPolicy::HandleStackOverflow() Line 147
	at D:\a\_work\1\s\src\coreclr\vm\eepolicy.cpp(147)
coreclr.dll!ProcessCLRException(_EXCEPTION_RECORD * pExceptionRecord, unsigned __int64 MemoryStackFp, _CONTEXT * pContextRecord, _DISPATCHER_CONTEXT * pDispatcherContext) Line 895
	at D:\a\_work\1\s\src\coreclr\vm\exceptionhandling.cpp(895)
ntdll.dll!RtlpExecuteHandlerForException() Line 132
	at minkernel\ntos\rtl\amd64\xcptmisc.asm(132)
ntdll.dll!RtlDispatchException(_EXCEPTION_RECORD * ExceptionRecord, _CONTEXT * ContextRecord) Line 595
	at minkernel\ntos\rtl\amd64\exdsptch.c(595)
ntdll.dll!KiUserExceptionDispatch() Line 751
	at minkernel\ntos\rtl\amd64\trampoln.asm(751)
00007fff9780041c()
        < Managed code >
```

### Regression?

This is a regression from .NET Framework.

### Known Workarounds

The only work around I have found is to run the code under the x86 runtime.

### Configuration

Version: .NET 6 and .NET 7
OS: Windows 10 (didn't try anything else)
Architecture: X64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by noahfalk**
In the function Debugger::IsThreadAtSafePlace we have some code that attempts to avoid calling a stack intensive helper method if the runtime is currently handling a StackOverflowException. The logic to do that check was incorrect because the runtime's EH has some special case handling for StackOverflow. Most exceptions cause the EH code to create an ExceptionTracker object which in turn causes thread->IsExceptionInProgress() to return TRUE. StackOverflow doesn't set that and goes straight to fatal error handling so thread->IsExceptionInProgress() is false.

The fix avoids checking IsExceptionInProgress(). g_pEEInterface->GetThreadException() works because it checks both the tracker and also falls back to thread->m_lastThrownObject where the StackOverflow exception will be stored.

Fixes https://github.com/dotnet/runtime/issues/70755

### CODE REVIEW

**Comment by noahfalk** [comment]
@janvorli @dotnet/dotnet-diag

**Review by tommcdon — APPROVED** [review]

**Review by hoyosjs — APPROVED** [review]

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

File: src/coreclr/debug/ee/debugger.cpp [hunk 1]
```diff
@@ -12595,21 +12595,16 @@ bool Debugger::IsThreadAtSafePlace(Thread *thread)
...
-    //
     // On the Stack Overflow code path calling IsThreadAtSafePlaceWorker as it is
-    // currently implemented is way too stack intensive. For now we cheat and just
-    // say that if a thread is in the middle of handling a SO it is NOT at a safe
-    // place. This is a reasonably safe assumption to make and hopefully shouldn't
-    // result in deadlocking the debugger.
-    if ( (thread->IsExceptionInProgress()) &&
-         (g_pEEInterface->GetThreadException(thread) == CLRException::GetPreallocatedStackOverflowExceptionHandle()) )
+    // currently implemented is way too stack intensive. Conservatively we assume that
+    // any thread handling an SO is not at a safe place.
```
Comment (mikelle-rogers) [inline src/coreclr/debug/ee/debugger.cpp:12600]:
```suggestion
    // any thread handling a SO is not at a safe place.
```

**Review by mikelle-rogers — COMMENTED** [review]
Nit

**Pull request merged**